File: Medium/Perfect Sum Problem/perfect-sum-problem.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution{

	public:
	int perfectSum(int arr[], int n, int total)
	{
        // Your code goes here
         const int modulo = 1e9 + 7;
       
 vector<vector<int>> dynamic_programming(n + 1, vector<int>(total + 1, 0));
 
    // Initialize the dp array
    dynamic_programming[0][0] = 1;
    for (int j = 1; j <= total; j++)
        dynamic_programming[0][j] = 0;
 
    // Fill the dp array
    for (int i = 1; i <= n; i++) {
        for (int j = 0; j <= total; j++) {
            if (arr[i - 1] <= j)
                dynamic_programming[i][j] = (dynamic_programming[i - 1][j] + dynamic_programming[i - 1][j - arr[i - 1]]) % modulo;
            else
                dynamic_programming[i][j] = dynamic_programming[i - 1][j];
        }
    }
 
    return dynamic_programming[n][total];
        
	}
	  
};

//{ Driver Code Starts.
int main() 
{
   	
   
   	int t;
    cin >> t;
    while (t--)
    {
        int n, sum;

        cin >> n >> sum;

        int a[n];
        for(int i = 0; i < n; i++)
        	cin >> a[i];

       

	    Solution ob;
	    cout << ob.perfectSum(a, n, sum) << "\n";
	     
    }
    return 0;
}

// } Driver Code Ends

File: Difficulty: Medium/Perfect Sum Problem/perfect-sum-problem.java
//{ Driver Code Starts
import java.io.*;
import java.util.*;

class Main {
    public static void main(String args[]) throws IOException {
        BufferedReader read = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(read.readLine().trim()); // Read number of test cases

        // Loop for each test case
        while (t-- > 0) {
            String line = read.readLine().trim(); // Read the array input
            String[] numsStr = line.split(" ");   // Split the input string by spaces
            int[] nums =
                new int[numsStr.length]; // Convert string array to integer array
            for (int i = 0; i < numsStr.length; i++) {
                nums[i] = Integer.parseInt(numsStr[i]);
            }

            int target = Integer.parseInt(read.readLine().trim()); // Read target sum

            Solution ob = new Solution(); // Create an object of Solution class
            System.out.println(
                ob.perfectSum(nums, target)); // Call perfectSum and print the result
            System.out.println("~");          // Call perfectSum and print the result
        }
    }
}
// } Driver Code Ends


class Solution {
    
    
    public int fun(int ind, int target, int[] nums, int[][] dp)
    {
        
        if(ind == nums.length)
        {
        if(target == 0) return 1;
        else return 0;
            
        }
        
        if(dp[ind][target] != -1) return dp[ind][target];
        
        int notpick = fun(ind + 1, target, nums, dp);
        int pick = 0;
        
        if(nums[ind] <= target)
        {
            pick = fun(ind + 1, target - nums[ind], nums, dp);
        }
        
        return dp[ind][target] = pick + notpick;
    }
    
    // Function to calculate the number of subsets with a given sum
    public int perfectSum(int[] nums, int target) {
        // code here
        
        int n = nums.length;
        
        
        int[][] dp = new int[n][target + 1];
        
        for(int i = 0; i < n; i++)
        {
            Arrays.fill(dp[i],-1);
        }
        return fun(0,target,nums,dp);
        
        
        
    }
}